lib/spack/spack/package.py

```diff
@@ -52,26 +55,127 @@
 import spack.repository
 import spack.url
 import spack.util.web
-
-from urlparse import urlparse
-from StringIO import StringIO
 from llnl.util.filesystem import *
 from llnl.util.lang import *
 from llnl.util.link_tree import LinkTree
 from llnl.util.tty.log import log_output
+from spack import directory_layout
 from spack.stage import Stage, ResourceStage, StageComposite
 from spack.util.compression import allowed_archive
 from spack.util.environment import dump_environment
-from spack.util.executable import ProcessError, Executable, which
+from spack.util.executable import ProcessError, which
 from spack.version import *
-from spack import directory_layout
-from urlparse import urlparse
 
 """Allowed URL schemes for spack packages."""
 _ALLOWED_URL_SCHEMES = ["http", "https", "ftp", "file", "git"]
 
 
-class Package(object):
+class InstallPhase(object):
+    """Manages a single phase of the installation
+
+    This descriptor stores at creation time the name of the method it should search
+    for execution. The method is retrieved at __get__ time, so that it can be overridden
+    by subclasses of whatever class declared the phases.
+
+    It also provides hooks to execute prerequisite and sanity checks.
+    """
+    def __init__(self, name):
+        self.name = name
+        self.preconditions = []
+        self.sanity_checks = []
```

Comment:
> So are we keeping the idea of sanity checks separate from tests? I agree that sanity checks should always be run, whether or not the user requests them with --run-tests or whatever flag we decide on.

How to name things for me is open to discussion. `preconditions` and `sanity_checks` are just two hooks to run code before and after a particular phase. The callbacks can be registered with a decorator as you can see in `AutotoolsPackage`. For instance `swiftsim` will fail before configure if you comment out the `autoreconf` method.

> We may also want to add a hook for filter_compilers.
￼

filter_compilers ? What do you mean ?